overlord/servicectl/servicectl.go

```diff
@@ -0,0 +1,95 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2015-2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package servicectl
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/overlord/cmdstate"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+type AppInstruction struct {
+	Action string   `json:"action"`
+	Names  []string `json:"names"`
+	client.StartOptions
+	client.StopOptions
+	client.RestartOptions
+}
+
+type UnknownAction struct{ error }
+type ServiceActionConflict struct{ error }
+
+func ServiceControl(st *state.State, appInfos []*snap.AppInfo, inst *AppInstruction) (*state.Change, error) {
+	// the argv to call systemctl will need at most one entry per appInfo,
+	// plus one for "systemctl", one for the action, and sometimes one for
+	// an option. That's a maximum of 3+len(appInfos).
+	argv := make([]string, 2, 3+len(appInfos))
+	argv[0] = "systemctl"
+
+	argv[1] = inst.Action
+	switch inst.Action {
+	case "start":
+		if inst.Enable {
+			argv[1] = "enable"
+			argv = append(argv, "--now")
+		}
+	case "stop":
+		if inst.Disable {
+			argv[1] = "disable"
+			argv = append(argv, "--now")
+		}
+	case "restart":
+		if inst.Reload {
+			argv[1] = "reload-or-restart"
+		}
+	default:
+		return nil, UnknownAction{fmt.Errorf("unknown action %q", inst.Action)}
```

Comment:
Error types are _always_ returned as pointers. If we break that rule, we will create very hard to debug errors because we won't be sure, will try to catch errors the wrong way, and unless there's explicit test coverage for the branch, which unfortunately isn't always true for errors, the code compiles fine and never hits.

With that said, why is that a special type?  In general we wait until we have a use case for the special type before introducing one.

Reply:
Ok, fixed. You're right, there is no need for special error type.